distributed/worker.py

```diff
@@ -1496,26 +1505,45 @@ def add_task(
             for dependency, workers in who_has.items():
                 assert workers
                 if dependency not in self.tasks:
+                    # initial state is "new"
+                    # this dependency does not already exist on worker
                     self.tasks[dependency] = dep_ts = TaskState(key=dependency)
-                    dep_ts.state = (
-                        "waiting" if dependency not in self.data else "memory"
-                    )
 
-                dep_ts = self.tasks[dependency]
-                self.log.append((dependency, "new-dep", dep_ts.state))
+                    # link up to child / parents
+                    ts.dependencies.add(dep_ts)
+                    dep_ts.dependents.add(ts)
 
-                if dep_ts.state != "memory":
-                    ts.waiting_for_data.add(dep_ts.key)
-                    self.waiting_for_data_count += 1
+                    # check to ensure task wasn't already executed and partially released
+                    # # TODO: make this less bad
+                    state = "fetch" if dependency not in self.data else "memory"
 
-                dep_ts.who_has.update(workers)
+                    # transition from new -> fetch handles adding dependency
+                    # to waiting_for_data
+                    self.transition(dep_ts, state)
 
-                ts.dependencies.add(dep_ts)
-                dep_ts.dependents.add(ts)
+                    self.log.append((dependency, "new-dep", dep_ts.state, ts))
 
-                for worker in workers:
-                    self.has_what[worker].add(dep_ts.key)
-                    if dep_ts.state != "memory":
+                else:
+                    # task was already present on worker
+                    dep_ts = self.tasks[dependency]
+
+                    # link up to child / parents
+                    ts.dependencies.add(dep_ts)
+                    dep_ts.dependents.add(ts)
+
+                    # possible that a new task has asked for a dependency we're already
+                    # planning on fetching -- make sure that task knows it has to
+                    # wait for its dependencies to arrive
+                    if dep_ts.state in ("fetch", "flight"):
+                        ts.waiting_for_data.add(dep_ts.key)
```

Comment:
IIRC, `waiting_for_data` to be empty is one of the conditions for the transition `waiting -> executing` and I'm wondering if this condition shouldn't also include `waiting` and `executing`

Reply:
I think that if I include `waiting` and `executing` in here that it might cause weird transitions -- if the dependency is `executing` and then we add it to `waiting_for_data` while it is moving to memory, it will cause us to try to re-run this and we'll end up with a `waiting -> memory` transition.

Reply:
sorry, I think my comment was wrong. This is an issue about semantics. I perceived `waiting_for_data` to be the same as "dependencies which are not yet in memory" but it is rather a "dependency which is not yet in memory and must be fetch from another worker". If this is true, your code changes make sense and I think there should be no other state in here